### DISCUSSION TITLE
Encode the CAN frames with dbcppp

### DISCUSSION THREAD

**Discussion by mounare** [xR3b0rn/dbcppp#72] [General]
Hi all,
I am using dbcppp with ROS to decode the CAN frames and it works fine. 
Now I want to encode the CAN frames, i.e, I have data and I want to assign them to different signals in the CAN FRAME, and then I send this frame to the CAN network
Is there an example to do this? @xR3b0rn  @derkomai  @ByunghunLee @byunghun-ee 
Thank you.

**Comment by xR3b0rn** [comment]
This can be done by using the `Signal`s `PhysToRaw` and `Encode` function:
```
inline void Encode(raw_t raw, void* buffer) const noexcept { return _encode(this, raw, buffer); }
inline raw_t PhysToRaw(double phys) const noexcept { return _phys_to_raw(this, phys); }
```
For example:
```
...
char buffer[8];
std::memset(buffer, 0, sizeof(buffer));
double phys_velocity = 100.;
auto raw_velocity = signal->PhysToRaw(phys_velocity);
signal->Encode(raw_velocity, reinterpret_cast<void*>(buffer));
```
In which the buffer must be able to keep at least as many bytes as the underlying message may has according to the parsed DBC.

**Comment by mounare** [comment]
Thank you for your reply.
I have tried but I think it does not work.
For example, suppose tI have the following CAN frame to encode (Just an example) with ID 0x100:
<img width="311" alt="fame_can" src="https://user-images.githubusercontent.com/87301173/125562054-b1f94ca7-6154-4d70-912d-8834c1ebdf4c.png">

To encode the first signal  "ID_of_Track" of the CAN frame, I did as following

```C++
// After reading DBC file, initializing node (ROS) and CAN Frame  
unsigned int track_ID=1;
dbcppp_msg->forEachSignal([&](const dbcppp::Signal& signal )
  {
    if ( signal.getName().compare("ID_of_Track") == 0  )
    {     
      auto rawID=    signal.physToRaw(track_ID);`
      signal.encode(rawC,&massage.data[0]);`
    }
  }); 
```
Any help please?

**Reply by xR3b0rn** [reply]
What is failing there, what is the value of `track_ID`, what is the value of `rawID`, what is the bit pattern of `message.data` before and after `encode` and which version of `dbcppp` you are using (e.g. commit hash)?

**Reply by mounare** [reply]
My goal is to encode the CAN frame with the definition presented in the previous figure. So, the idea is to define a CAN frame and after that, I encode or assign the values for each signal. The values of these signals are determined from other nodes of ROS or sensors. 
The following table presents more details about the definition of CAN Frame (Note that Track means an obstacle detected):
<img width="516" alt="def" src="https://user-images.githubusercontent.com/87301173/125701182-52633c9a-ba90-44ec-bd65-0654f5b58e6d.png">
The class means the type of obstacles like a car or bike... and represented by an unsigned int (1 for car, 2 for Bike, 3 for bus ...). The Id is the identity of the obstacle ( unsigned int ).

I mention the following example, the signals of a CAN frame, that will be sent, have the following values:

<img width="149" alt="exm" src="https://user-images.githubusercontent.com/87301173/125702195-8353e753-867f-4bf3-86d2-08d15e3bd4dd.png">

For the C++ code, I tried the following program :
```
// After reading DBC file, initializing node (ROS) 
// Declaration of  Can Frame the will be encoded and sent
    can_msgs::Frame message;
    message.is_extended = false;
    message.is_rtr = false;
    message.is_error = false;
    message.dlc = 8;
    message.id = 0x1FF;
    message.header.frame_id = "0";  
    message.header.stamp = ros::Time::now();

unsigned int track_ID_Value=12;
unsigned int Class_of_Track_Value=1;
float Track_PosLat_Value=3,4 ;
// ...... Same for the rest of signals 
dbcppp_msg->forEachSignal([&](const dbcppp::Signal& signal )
  {
    if ( signal.getName().compare("ID_of_Track") == 0  )
    {     
      auto rawID=    signal.physToRaw(track_ID_Value);`
      signal.encode(rawID,&message.data[0]);`
    }

 if ( signal.getName().compare("Class_of_Track") == 0  )
    {     
      auto rawClass=    signal.physToRaw(Class_of_Track_Value);
      signal.encode(rawClass,&message.data[0]);`
    }

 if ( signal.getName().compare("Track_PosLat") == 0  )
    {     
      auto rawTrack_PosLat=    signal.physToRaw(Track_PosLat_Value);
      signal.encode(rawTrack_PosLat,&message.data[0]);`
    }

// ...... Same for the rest of the signals 

  }); 
```


So, it is right to encode the CAN frame like this with dbcppp?
Thank you.

**Comment by xR3b0rn** [comment]
Your code looks fine, though I recommend to not use the dbcppp's DBC API for the lookup. The dbcppp is only meant to be a parser. For lookup, you better go with a hash map or a vector in order to access the desired signal with better performance. For an example you can look at the message usage in the [BasicExample](https://github.com/xR3b0rn/dbcppp/blob/2c2eed134626783f0108824224ac5c2350b4b72d/examples/BasicUsage/main.cpp#L58).

If you want more specific help, you have either update to the current version of [dbcppp](https://github.com/xR3b0rn/dbcppp/releases/tag/v3.0.1-stable-api) or provide a reference to the version you use. Additionally a small program which reprdocues the error you are reffering to would be of great help.
